src/oauth2/service.ts

```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
 
 }
 
+export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) {
+  if (!codeChallenge && !codeVerifier) {
+    // This request was not initiated with PKCE support, so ignore the validation
+    return;
+  }
+
+  if (codeChallenge && !codeVerifier) {
+    // The authorization request started with PKCE, but the token request did not follow through
+    throw new InvalidRequest('The code verifier was not supplied');
+  }
+
+  // For the plain method, the derived code and the code verifier are the same
+  let derivedCodeChallenge = codeVerifier;
```

Comment:
Is `plain` secure? I don't really know the PKCE spec too well!

Reply:
No, it's to accommodate clients that cannot do sha256 for technical reasons since PKCE is required in 2.1.

Reply:
I see. I remember one of the major issues with JWT was that many libraries by default also supported a 'none' encryption method. So I hope this is not a similar thing, but I would guess not.

Reply:
I would hope oauth libraries that support sha256 wouldn't provide the implementor the option of disabling it. The RFC does say the client MUST use sha256 if it can technically.
https://tools.ietf.org/html/draft-parecki-oauth-v2-1-01#section-4.1.1.2